### PR TITLE
Phase 6: Charbonnier Loss — Fully Smooth L1 for Dense Pressure Regression

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current training loss uses L1 (mean absolute error): `abs_err = (pred - y_norm).abs()`. L1 has a non-differentiable kink at zero — the gradient is ±1 everywhere except exactly at 0. For a well-trained model operating in a narrow residual band, this creates a plateau: the sign of the gradient is correct but the magnitude carries no magnitude information about how far off the prediction is.

**Charbonnier loss** replaces this with:
```
abs_err = sqrt((pred - y_norm)^2 + eps^2) - eps
```

This is fully smooth (C∞) everywhere. The gradient is `(pred - y_norm) / sqrt((pred - y_norm)^2 + eps^2)`, which interpolates between near-zero for small residuals and ±1 for large residuals. The transition is controlled by `eps`.

**Physical motivation:** Surface pressure prediction has a non-Gaussian error distribution — there are sharp suction peaks (large errors) and flat pressure recovery regions (small errors). The smooth gradient of Charbonnier provides proportional signal across this range, whereas L1 treats a 0.01 residual the same as a 1.0 residual (both gradient = ±1). For the surface refinement head (which operates on already-small residuals), the kink-free gradient may provide more consistent optimization signal.

This is distinct from Huber loss (#2113, edward) which has a hard threshold beta below/above which it switches between L2 and L1. Charbonnier has no threshold — the gradient is continuous and smooth everywhere, asymptotically approaching L1 for large residuals.

Reference: Barron (2019) "A General and Adaptive Robust Loss Function", CVPR — Charbonnier is a known-good special case of this family for dense regression tasks with non-Gaussian error distributions.

## Instructions

**Step 1: Add the Charbonnier eps flag** to argparse:
```python
parser.add_argument('--charbonnier_eps', type=float, default=0.0,
                    help='Charbonnier loss eps (0=disabled, uses standard L1)')
```
Add to Config dataclass:
```python
charbonnier_eps: float = 0.0
```

**Step 2: Find the L1 loss computation** in `train.py`. Look for `abs_err = (pred - y_norm).abs()` or similar — the per-node per-channel absolute error that feeds into vol_loss, surf_loss, and any hard-node mining.

**Step 3: Replace with Charbonnier when enabled:**
```python
if cfg.charbonnier_eps > 0:
    abs_err = ((pred - y_norm) ** 2 + cfg.charbonnier_eps ** 2).sqrt() - cfg.charbonnier_eps
else:
    abs_err = (pred - y_norm).abs()
```

That is the **complete change** — `abs_err` flows into all downstream loss terms (vol_loss, surf_loss, coarse pooling, hard-node mining) unchanged. The Charbonnier replaces L1 everywhere the base absolute error is used.

**Step 4: Verify correctness.** Confirm that when `charbonnier_eps=0`, output matches `abs()` numerically. Confirm that when eps=0.1 and pred=y_norm, abs_err=0 (the `-eps` cancels `sqrt(eps^2)=eps`).

**Step 5: Sweep eps values.** Run 2 seeds (42, 43) × 3 eps values = 6 runs total, all in parallel:
```bash
BASE="--agent alphonse --wandb_group phase6/charbonnier-loss \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3"

# eps=0.05 — tight smooth zone, closest to L1
python train.py $BASE --charbonnier_eps 0.05 --seed 42 --wandb_name "alphonse/charb-eps005-s42"
python train.py $BASE --charbonnier_eps 0.05 --seed 43 --wandb_name "alphonse/charb-eps005-s43"

# eps=0.1 — standard from image restoration literature
python train.py $BASE --charbonnier_eps 0.1  --seed 42 --wandb_name "alphonse/charb-eps01-s42"
python train.py $BASE --charbonnier_eps 0.1  --seed 43 --wandb_name "alphonse/charb-eps01-s43"

# eps=0.2 — wider smooth zone, more L2-like near zero
python train.py $BASE --charbonnier_eps 0.2  --seed 42 --wandb_name "alphonse/charb-eps02-s42"
python train.py $BASE --charbonnier_eps 0.2  --seed 43 --wandb_name "alphonse/charb-eps02-s43"
```

**What to report:** Table of (eps, seed, p_in, p_oodc, p_tan, p_re, val/loss, W&B run ID) + seed-averaged means per eps. Compare against the 8-seed single-model mean baseline (p_in=13.03, p_oodc=7.83, p_tan=30.29, p_re=6.45).

## Baseline

Current best (16-seed ensemble, PR #2093):

| Metric | 16-Seed Ensemble | 8-Seed Single-Model Mean |
|--------|-----------------|--------------------------|
| p_in   | **12.1**        | 13.03                    |
| p_oodc | **6.6**         | 7.83                     |
| p_tan  | **29.1**        | 30.29                    |
| p_re   | **5.8**         | 6.45                     |

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent alphonse \
  --wandb_name "alphonse/baseline-s42" --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results — Charbonnier Loss

All 8 runs completed at epoch 157 (180-min wall-clock timeout). W&B group: `phase6/charbonnier-loss`.

### Results Table

| Config | Seed | p_in | p_oodc | p_tan | p_re | val/loss | W&B Run ID |
|--------|------|------|--------|-------|------|----------|------------|
| **Baseline (L1)** | 42 | **12.87** | **7.9** | **30.4** | **6.5** | 0.3906 | `wemzzic2` |
| **Baseline (L1)** | 43 | **12.59** | **8.0** | **29.6** | **6.5** | 0.3851 | `v7ulm4nt` |
| Charb eps=0.05 | 42 | 14.17 | 8.7 | 30.7 | 7.2 | 0.4054 | `d79f5gab` |
| Charb eps=0.05 | 43 | 13.54 | 9.1 | 30.7 | 7.2 | 0.4030 | `r619gn21` |
| Charb eps=0.1 | 42 | 13.87 | 9.4 | 30.5 | 7.2 | 0.4056 | `51vbivxc` |
| Charb eps=0.1 | 43 | 14.03 | 9.6 | 30.8 | 7.2 | 0.4095 | `4ert76ho` |
| Charb eps=0.2 | 42 | 14.72 | 9.6 | 30.3 | 7.5 | 0.4160 | `br80gpr2` |
| Charb eps=0.2 | 43 | 14.58 | 10.1 | 32.2 | 7.8 | 0.4234 | `290ls7cp` |

### Seed-Averaged Means

| Config | p_in | p_oodc | p_tan | p_re | val/loss |
|--------|------|--------|-------|------|----------|
| **Baseline (L1)** | **12.73** | **7.95** | **30.00** | **6.50** | **0.3879** |
| Charb eps=0.05 | 13.85 (+8.8%) | 8.90 (+12.0%) | 30.70 (+2.3%) | 7.20 (+10.8%) | 0.4042 |
| Charb eps=0.1 | 13.95 (+9.6%) | 9.50 (+19.5%) | 30.65 (+2.2%) | 7.20 (+10.8%) | 0.4076 |
| Charb eps=0.2 | 14.65 (+15.1%) | 9.85 (+23.9%) | 31.25 (+4.2%) | 7.65 (+17.7%) | 0.4197 |

**8-seed single-model baseline for reference:** p_in=13.03, p_oodc=7.83, p_tan=30.29, p_re=6.45

### Analysis

**Negative result.** Charbonnier loss degrades all surface MAE metrics at every tested epsilon value.

Key observations:
1. **Monotonic degradation with eps**: Larger eps = worse performance across all metrics. eps=0.2 is worst (p_oodc +24% vs baseline).
2. **p_oodc hit hardest**: OOD-condition generalization suffers most (+12% to +24%), suggesting the smooth gradient near zero hurts the model's ability to learn sharp pressure features needed for unseen conditions.
3. **p_tan relatively stable**: Tandem transfer is least affected (+2-4%), likely because tandem errors are dominated by large residuals where Charbonnier ≈ L1.
4. **p_re consistently worse**: +11-18% degradation across all eps values.
5. **Even eps=0.05 (closest to L1) hurts**: The smooth zone doesn't need to be large to cause damage — the gradient attenuation near zero removes important optimization signal for small residuals.

**Why it fails**: The hypothesis was that the L1 kink at zero creates a gradient plateau for small residuals. In practice, the ±1 gradient of L1 is actually beneficial — it provides constant optimization pressure on all residuals regardless of magnitude. The Charbonnier's gradient attenuation near zero (~proportional to residual magnitude) slows convergence on the many near-zero residuals that dominate the well-predicted bulk of the field. The model spends gradient budget on large errors (same as L1) but underweights the fine-tuning of small errors that matter for surface MAE.

### Suggested Follow-ups

- **Reverse Charbonnier / anti-smooth loss**: Instead of smoothing near zero, consider a loss that provides *stronger* gradient for small residuals (super-linear near zero). E.g., `|x|^p` with p < 1, or a piecewise loss that upweights small errors.
- **Asymmetric loss**: Pressure under-prediction vs over-prediction may have different physical consequences. An asymmetric penalty could improve surface Cp accuracy.
- **Focal-style reweighting**: Instead of changing the loss shape, modulate per-sample or per-node weights based on current error magnitude to focus on hard examples.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2116 because **Dead end — closing.** Charbonnier loss degrades all surface MAE metrics at every tested epsilon value:

| Config | p_in | p_oodc | p_tan | p_re |
|--------|------|--------|-------|------|
| Baseline (L1) | 12.73 | 7.95 | 30.00 | 6.50 |
| Charb eps=0.05 | 13.85 (+8.8%) | 8.90 (+12.0%) | 30.70 (+2.3%) | 7.20 (+10.8%) |
| Charb eps=0.1 | 13.95 (+9.6%) | 9.50 (+19.5%) | 30.65 (+2.2%) | 7.20 (+10.8%) |
| Charb eps=0.2 | 14.65 (+15.1%) | 9.85 (+23.9%) | 31.25 (+4.2%) | 7.65 (+17.7%) |

Even the smallest eps (0.05, closest to L1) hurts. The student's analysis is correct: L1's constant ±1 gradient is actually a feature, not a bug — it provides uniform optimization pressure on all residuals. Charbonnier's gradient attenuation near zero slows convergence on small residuals that dominate near-converged surface MAE. The smooth loss family (Huber in #2113, Charbonnier here) has now been thoroughly explored and consistently underperforms L1. The L1 loss shape is well-suited to this problem.

The student's suggested 'Reverse Charbonnier / super-linear near zero' direction is interesting and worth exploring in a future assignment..
